### PR TITLE
Move hermes confs to zookeeper

### PR DESCRIPTION
**Pull request by jeanleonov**
- Move hemres configs to zookeeper (AppController writes configs to zookeeper, Hermes watches zookeeper nodes and applies changes without restart).
 - Handle error when collecting cluster stats and report failures.

It also includes changes from https://github.com/AppScale/appscale/pull/2466:
 - Profiling interval become configurable in runtime.
 - Hermes writes fewer logs.
 - Fixed bug with getting UTC timestamp for profile log.

### CODE REVIEW

**Comment by jeanleonov** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/3504/

File: Hermes/appscale/hermes/stats/stats_app.py [hunk 1]
```diff
@@ -144,6 +154,98 @@ def get_cluster_stats_api_routes(is_master):
...
+      PROXIES_STATS_CONFIGS_NODE, self.update_proxies_profiling_conf
+    )
+
+  def update_nodes_profiling_conf(self, new_conf, znode_stat):
+    if not new_conf:
+      logging.debug("No node stats profiling configs are specified yet")
+      return
+    logging.info("New nodes stats profiling configs: {}".format(new_conf))
+    conf = json.loads(new_conf)
+    enabled = conf.get("enabled")
```
Comment (cdonati) [inline Hermes/appscale/hermes/stats/stats_app.py]: Should this default to False so you can treat enabled as a boolean?

File: Hermes/appscale/hermes/stats/stats_app.py [hunk 2]
```diff
@@ -144,6 +154,98 @@ def get_cluster_stats_api_routes(is_master):
...
 
+class ProfilingManager(object):
+  def __init__(self, zk_client):
+    self.nodes_profile_log = None
+    self.processes_profile_log = None
+    self.proxies_profile_log = None
+    self.nodes_profile_task = None
+    self.processes_profile_task = None
+    self.proxies_profile_task = None
+    zk_client.DataWatch(
```
Comment (cdonati) [inline Hermes/appscale/hermes/stats/stats_app.py]: These watches run outside the IO loop in a separate thread. To make things simpler, I would recommend having the watch functions bridge back to the main IO loop (with `IOLoop.instance().add_callback`, for example) and do any state mutations and callbacks from there.

That way, you don't need to worry about if you need a threading lock around certain reads and mutations or if your periodic callbacks are in different threads.

**Comment by jeanleonov** [comment]
Run failed tests again, all blue:
https://ocd.appscale.com:8080/job/Hawkeye-Advanced-8/512/
https://ocd.appscale.com:8080/job/Hawkeye-Simple-8/562/
https://ocd.appscale.com:8080/job/Hawkeye/11401/

File: Hermes/appscale/hermes/stats/stats_app.py (hunk 2 above)
Comment (jeanleonov) [inline Hermes/appscale/hermes/stats/stats_app.py]: Thanks! I missed that.

**Comment by jeanleonov** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/3586/
Still blue after resolving merge conflicts.

File: Hermes/appscale/hermes/stats/stats_app.py [hunk 3]
```diff
@@ -144,6 +154,101 @@ def get_cluster_stats_api_routes(is_master):
...
+    self.nodes_profile_log = None
+    self.processes_profile_log = None
+    self.proxies_profile_log = None
+    self.nodes_profile_task = None
+    self.processes_profile_task = None
+    self.proxies_profile_task = None
+
+    def bridge_to_ioloop(update_function):
+      def do_in_ioloop():
+        IOLoop.current().add_callback(update_function)
```
Comment (cdonati) [inline Hermes/appscale/hermes/stats/stats_app.py]: Doesn't this need to pass in the arguments, or is there some magic happening here that I'm missing?

File: Hermes/appscale/hermes/stats/stats_app.py [hunk 4]
```diff
@@ -144,6 +154,101 @@ def get_cluster_stats_api_routes(is_master):
   ]
 
 
+class ProfilingManager(object):
```
Comment (cdonati) [inline Hermes/appscale/hermes/stats/stats_app.py]: I think a few docstrings would be nice for consistency.

File: Hermes/appscale/hermes/stats/stats_app.py (hunk 3 above)
Comment (jeanleonov) [inline Hermes/appscale/hermes/stats/stats_app.py]: Thanks for noticing! No magic here, I just lost a commit which fixes this problem in a different branch=(

**Comment by jeanleonov** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/3600/
All blue

File: Hermes/appscale/hermes/stats/stats_app.py (hunk 4 above)
Comment (jeanleonov) [inline Hermes/appscale/hermes/stats/stats_app.py]: done

File: Hermes/appscale/hermes/stats/stats_app.py [hunk 5]
```diff
@@ -144,6 +154,143 @@ def get_cluster_stats_api_routes(is_master):
...
+      """ Creates function which schedule execution of update_function
+      inside current IOLoop.
+
+      Args:
+        update_function: a function to execute in IOLoop.
+      Returns:
+        A callable which schedules execution of update_function inside IOLoop.
+      """
+      def update_in_ioloop(new_conf, znode_stat):
+        update = lambda: update_function(new_conf, znode_stat)
```
Comment (cdonati) [inline Hermes/appscale/hermes/stats/stats_app.py]: No need to change it now, but just for future reference, `add_callback` accepts arguments so the lambda is not necessary.

**Comment by jeanleonov** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/3678/ - in progress

**Comment by jeanleonov** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/3689/
This one looks better

**Pull request merged**